dbms/src/Functions/geohashCoverAreaWithBoxes.cpp

```diff
@@ -0,0 +1,172 @@
+                    const IColumn * precision_column,
+                    ColumnPtr & result)
+    {
+        static constexpr size_t max_array_size = 10'000'000;
+
+        const auto * lon_min = checkAndGetColumn<ColumnVector<LonAndLatType>>(lon_min_column);
+        const auto * lat_min = checkAndGetColumn<ColumnVector<LonAndLatType>>(lat_min_column);
+        const auto * lon_max = checkAndGetColumn<ColumnVector<LonAndLatType>>(lon_max_column);
+        const auto * lat_max = checkAndGetColumn<ColumnVector<LonAndLatType>>(lat_max_column);
+        auto * precision = checkAndGetColumn<ColumnVector<PrecisionType>>(precision_column);
+        if (precision == nullptr)
+        {
+            precision = checkAndGetColumnConstData<ColumnVector<PrecisionType>>(precision_column);
+        }
+
+        if (!lon_min || !lat_min || !lon_max || !lat_max || !precision)
+            return false;
+
+        const size_t total_rows = lat_min->size();
+
+        auto col_res = ColumnArray::create(ColumnString::create());
+        ColumnString & res_strings = typeid_cast<ColumnString &>(col_res->getData());
+        ColumnArray::Offsets & res_offsets = col_res->getOffsets();
+        ColumnString::Chars & res_strings_chars = res_strings.getChars();
+        ColumnString::Offsets & res_strings_offsets = res_strings.getOffsets();
+
+        for (size_t row = 0; row < total_rows; ++row)
+        {
+            const Float64 lon_min_value = lon_min->getElement(row);
+            const Float64 lat_min_value = lat_min->getElement(row);
+            const Float64 lon_max_value = lon_max->getElement(row);
+            const Float64 lat_max_value = lat_max->getElement(row);
+            const PrecisionType precision_value = GeoUtils::geohashPrecision(precision->getElement(row % precision->size()));
+
+            const auto estimated_array_size = GeoUtils::geohashEstimateCoverAreaWithBoxesItems(
+                        lon_min_value, lat_min_value, lon_max_value, lat_max_value, precision_value);
+
+            if (estimated_array_size > max_array_size)
+            {
+                throw Exception{getName() + " would produce " + std::to_string(estimated_array_size) +
+                    " array elements, which is greater than the allowed maximum of " + std::to_string(max_array_size),
+                    ErrorCodes::ARGUMENT_OUT_OF_BOUND};
+            }
+
+            res_strings_offsets.reserve(res_strings_offsets.size() + estimated_array_size);
+            res_strings_chars.resize(res_strings_chars.size() + estimated_array_size * (precision_value + 1));
+            const auto starting_offset = res_strings_offsets.empty() ? 0 : res_strings_offsets.back();
+            char * out = reinterpret_cast<char *>(res_strings_chars.data() + starting_offset);
+
+            const auto actual_array_size = GeoUtils::geohashCoverAreaWithBoxes(
+                        lon_min_value, lat_min_value, lon_max_value, lat_max_value, precision_value, out);
```

Comment:
Can you just throw in geohashCover.... if you have too many boxes? You won't need all the other functions then. Or maybe even code all this in place. Having three embedded loops is slightly inconvenient, but they are very simple so it's probably ok.

Reply:
I don't see why, most of the code here is regarding forcing data into Array of String-s, so removing a tiny bit that throws an exception wouldn't bring much simplicity.

* `geohashPrecision` is required to properly estimate string storage and properly compute string offsets.
* `geohashEstimateCoverAreaWithBoxesItems` is required to estimate string storage too.

Reply:
I mean the `prologue` function. Apparently it calculates the number of small boxes -- couldn't tell from the name. It returns a three field tuple -- I already forgot which field is which. You call it twice -- first, through a single-line wrapper (does the wrapper really add anything?) to get the size of resulting array. Then it is called again -- now by the function that creates the boxes. So the calculation is done twice (well ok it's fast), and two places in code have the assumption that they do the same calculation (again, doesn't matter here but in general that's ugly). If you have to split this in several functions, at least make a struct for the state of the algorithm, that has properly named fields, and calculate them once. You can put input parameters there as well.

Reply:
Saved one computation, made named struct, reduced number of geohash-related functions in `GeoUtils`